add ocs api for public link share email notification

karakayasemi opened the pull request:
## Description
This PR adds an OCS API for public link e-mail notifications. old ajax endpoints removed and oc10 adjusted to use the new API.

I aimed to keep the same functionality as oc10. In addition, the current code is simplified by removing redundant post parameters:
- The old endpoint is taking many pieces of information related to the share of link as POST parameter like expiration date, filename, itemType, etc. It is also trying to validate them on the backend. However, we can already extract these parameters by reaching related share object on the backend. So, I considered them redundant and, removed.
- I could not find any usage of `cc` and `bcc` fields in the codebase. So, I removed these options regards to YAGNI principle.
- When `toSelf` parameter is posted, the backend side is only adding user's e-mail to the recipient list. I moved this logic to the client-side and removed `toSelf` from POST parameters.

### UPDATE
- Internal email notification mechanism also moved to the new controller.

## Usage
* **URL**
/ocs/v1.php/apps/files_sharing/api/v1/notification/notifypubliclinkwithemail

* **Method:**
  `POST`
  
*  **POST Params**
   **Required:**
   `recipients=[string[]]`
   `link=[string]`
   `personalNote=[string]`
```
* **Success Response:**
{"ocs":{"meta":{"status":"ok","statuscode":100,"message":"OK","totalitems":"","itemsperpage":""},"data":[]}}
 
* **Fail Response:**
{"ocs":{"meta":{"status":"error","statuscode":400,"message":"Couldn't send mail to following recipient(s): test@email.com","totalitems":"","itemsperpage":""},"data":[]}}
```
## Related Issue
- Part of #35983

## Motivation and Context
In Phoenix, shared links shall be sent via email. We need an API for that.

## How Has This Been Tested?
- Unit tests are added.
- Manually by sending public link e-mails from the web interface.
- Manually with API call.

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Database schema changes (next release will require increase of minor version instead of patch)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [ ] Tests only (no source changes)

## Checklist:
<!-- Tick the checkboxes when done. -->
<!-- Raise documentation ticket in https://github.com/owncloud/documentation -->
- [x] Code changes
- [x] Unit tests added
- [ ] Acceptance tests added
- [ ] Documentation ticket raised: <link>

karakayasemi committed:
add ocs api for public link share email notification

DeepDiver1975 commented on apps/files_sharing/appinfo/routes.php:
```diff
@@ -96,6 +96,11 @@
 		],
+		[
+			'name' => 'Notification#notifyPublicLinkRecipientsWithEmail',
+			'url' => '/api/v1/notification/notifypubliclinkwithemail',
```
maybe kebab style looks better? notify-public-link-by-email
Or pascal notifyPublicLinkByEmail

karakayasemi replied:
now it is kebab style.

DeepDiver1975 commented on apps/files_sharing/lib/Controller/NotificationController.php:
```diff
@@ -0,0 +1,119 @@
+use OCP\Share\IManager;
+use Symfony\Component\EventDispatcher\EventDispatcherInterface;
+
+/**
```
please move to the beginning of the file

karakayasemi replied:
resolved

DeepDiver1975 commented on apps/files_sharing/lib/Controller/NotificationController.php:
```diff
@@ -0,0 +1,119 @@
+	public function notifyPublicLinkRecipientsWithEmail($link, $recipients, $personalNote) {
+		$message = null;
+		$code = 100;
+		$mailNotification = new MailNotifications(
```
maybe simplify this be injecting class MailNotifications?

karakayasemi replied:
If we refactor this class, it will be unit testable. Since Internal email notification part of the code is also using this class. I can move internal email notification to the new controller after refactoring.

karakayasemi replied:
`MailNotifications` class is now injectable. Internal sharing email notification mechanism also moved to the new controller. PR description updated to according to this.

DeepDiver1975 commented on apps/files_sharing/lib/Controller/NotificationController.php:
```diff
@@ -0,0 +1,119 @@
+	 * @NoAdminRequired
+	 *
+	 * @param string $link
+	 * @param string $recipients
```
string[] ??

DeepDiver1975 replied:
also add type hint array for this argument

karakayasemi replied:
its type is string. if it has multiple emails, they are separated with a comma. `MailNotifications` class is accepting input in this way. If you prefer in array type, I can refactor it also.

DeepDiver1975 replied:
Please at least add a comment then - THX a lot

karakayasemi replied:
After refactoring, now it takes `string[]`. Doc updated.

DeepDiver1975 commented on apps/files_sharing/lib/Controller/NotificationController.php:
```diff
@@ -0,0 +1,119 @@
+	/**
+	 * @NoAdminRequired
+	 *
+	 * @param string $link
```
~This way this API could be used for sending any data to anybody. We might want to change the api to reference a share id and also check if the current use is allowed to access the share.~

DeepDiver1975 commented:
- [ ] The email subject did change. Before this change: 'Admin shared foo.txt with you' After the change: 'Admin shares public link with you' - link name va file name
- [ ] ^ same with the email body
- [ ] unrelated to this change but found while testing: the subject is html escaped - which is bad
![Screenshot from 2019-08-19 16-50-31](https://user-images.githubusercontent.com/1005065/63275290-919cb400-c2a1-11e9-8232-c6bddcc18de1.png)

DeepDiver1975 committed:
Don't escape the file name in the email subject

DeepDiver1975 commented:
- [ ] acceptance tests need to be adjusted I guess - see https://drone.owncloud.com/owncloud/core/20273/76/14

karakayasemi committed:
make MailNotifications injectable, move internal mail notifications to new controller

karakayasemi commented:
@DeepDiver1975 Reason behind the e-mail subject and, body change is related to my mistake. I mistakenly used `$share->getName()` as file name. Now, I corrected it. Html escape issue is also related to this mistake. Since it's no longer needed, I've reverted your last commit changes.

Acceptance test fail is also related to this mistake. I hope this time, it will pass. Please, review the PR one more time. Thanks.

DeepDiver1975 commented:
> Html escape issue is also related to this mistake. Since it's no longer needed, I've reverted your last commit changes.

If the file name holds e.g. äöü - the subject is not escaped? This issue also existed before this PR - please double check. THX

jvillafanez commented on apps/files_sharing/lib/Controller/NotificationController.php:
```diff
@@ -0,0 +1,169 @@
+
+		$userFolder = $this->rootFolder->getUserFolder($sender->getUID());
+		$nodes = $userFolder->getById($itemSource, true);
+		$node = $nodes[0] ?? null;
```
`$node = $nodes[0]` ? I don't think we need more than that...
In addition, what happens with the rest of the nodes? Consider to add a comment explainig.

karakayasemi replied:
@jvillafanez in the beginning, this pr aimed only to add ocs api for public link email notification. However, I had to refactor `MailNotifications` class to make it injectable and testable. 
The only other consumer of this class was this method.  Since `Share20OcsController` already pretty big, that's why I did not want to inject new dependency to it. To keep the change minimum, I just moved this method from `Share20OcsController` to the new Notification Controller and, adapted the code to use injected `MailNotifications` class. So, I did not interest with the logic behind this method. I guess this node will be used for checking both sender and recipient have this node  :confused:.

jvillafanez replied:
we'll have to check it at some point. It's ok for now.

settermjd commented:
@karakayasemi, I've not been able to use the API successfully, as yet. The only response I get is the following:

```xml
<?xml version="1.0"?>
<ocs>
 <meta>
  <status>failure</status>
  <statuscode>403</statuscode>
  <message>Public link mail notification is not allowed</message>
  <totalitems></totalitems>
  <itemsperpage></itemsperpage>
 </meta>
 <data/>
</ocs>
```

That said, the following two configuration options are in effect:

```php
'shareapi_allow_public_notification' => 'yes',
'dav.enable.tech_preview' => true,
```

Any suggestions?

karakayasemi commented:
![image](https://user-images.githubusercontent.com/14157973/66209444-0431e980-e6c0-11e9-90a3-89e8fd4f0875.png)

Please enable the option that I highlighted in the image from admin sharing settings.

settermjd commented:
Thanks for the clarification. I had thought that setting the config value would be sufficient.

settermjd commented:
It seems I don't have the same version as you do. 

![Screenshot_2019-10-04 Settings - ownCloud](https://user-images.githubusercontent.com/196801/66210239-9f739080-e6b9-11e9-82d7-190097203e1a.png)

settermjd commented:
I'm running 10.3.0 alpha2 (cloned from master).

karakayasemi commented:
There is another option appearing after allowing users to share via link.
